### PR TITLE
Add Filter Restrictions handler

### PR DESCRIPTION
**Pull request by adrian-musora**
Design:
- queries that require modifications on the filter that were previously handled by ContentService/ContentReposity now can use a `FilterBuilder` instance to handle the creation of the filter arguments. 
- Added ulitity function to build queries raw, or using the filter builder option
- clean up existing methods to use utility functions or explicitly call the generic version of the method
- add (easy) missing tests to check that methods generate valid groq. The tests don't fully test the functionality of the methods, only compilation of groq.

Further Work to do:
- `fetchAll` was not touched. This will require it's own PR to complete because of the complexity. I wanted to get the framework in place to avoid too many merge conflicts/updates. 
- see table below. Some methods need to be refactored or updated

Testing information (what I did)

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>
<head>	
</head>

<body>

Method | Uses New Filter | User Raw Query | Hardcoded | Uses FetchAll | Has Test |  
-- | -- | -- | -- | -- | -- | --
fetchSongById | 1 |   |   |   | 1 |  
fetchArtists | 1 |   |   |   | 1 | This maybe shouldn’t. Not sure what this is used for
fetchSongArtistCount | 1 |   | 1 |   | 1 |  
fetchRelatedSongs |   |   | 1 |   | X | This needs to be refactored
fetchAllSongs |   |   |   | 1 | 1 |  
fetchSongFilterOptions |   | 1 |   |   | 1 | Should be moved to filter I think
fetchSongCount |   |   | 1 |   | 1 |  
fetchWorkouts | 1 |   |   |   | 1 |  
fetchNewReleases | 1 |   |   |   | 1 |  
fetchUpcomingEvents | 1 |   |   |   | 1 |  
fetchByRailContentId |   | 1 |   |   | 1 |  
fetchByRailContentIds |   |   | 1 |   | 1 |  
fetchAll |   |   |   | 1 | 1 | several tests
fetchAllFilterOptions |   |   | 1 |   | 1 | Should be moved to filter I think
fetchFoundation | 1 |   |   |   | 1 |  
fetchMethods |   |   | 1 |   | 1 |  
fetchMethod |   |   | 1 |   | 1 | Should be moved to filter I think
fetchMethodChildren |   |   |   |   | X |  
fetchMethodNextLesson |   |   |   |   | X |  
fetchMethodChildrenIds |   |   |   |   | X |  
fetchNextPreviousLesson |   |   |   |   | X |  
fetchRelatedLessons |   |   | 1 |   | 1 | Should be moved to filter I think
fetchAllPacks | 1 |   |   |   | 1 |  
fetchPackAll | 1 |   |   |   | 1 |  
fetchPackChildren |   |   | 1 |   | X |  
fetchLessonContent | 1 |   |   |   | 1 | Maybe shouldn’t be?
fetchCourseOverview | 1 |   |   |   | 1 |  
fetchChildren |   |   | 1 |   | 1 |  
fetchParentByRailContentId |   |   | 1 |   | 1 |  
fetchMethodPreviousNextLesson |   |   |   |   | X |  
fetchChallengeOverview |   |   | 1 |   | 1 |  
fetchCoachLessons |   | 1 |   |   | 1 |  
fetchByReference |   | 1 |   |   | 1 |  

</body>

</html>

### CODE REVIEW

File: src/filterBuilder.js
Comment (adrian-musora) [inline src/filterBuilder.js:122]:
```suggestion
        while( this.filter.charAt(0) === '&'  || this.filter.charAt(0) === ' ' ) this.filter = this.filter.substring(1);
        while( this.filter.charAt(this.filter.length) === '&' || this.filter.charAt(this.filter.length) === ' ' ) this.filter = this.filter.slice(-1);
```

File: src/services/sanity.js
Comment (BrandonKerr) [inline src/services/sanity.js:287]: Shouldn't these use the variables?
```suggestion
          sortOrder: sortOrder,
          end: end,
```

File: src/services/sanity.js
Comment (BrandonKerr) [inline src/services/sanity.js:380]: Really nit picky, but there's an unnecessary space
```suggestion
* @param {string} contentType - The document type of content to fetch
```

File: test/sanityQueryService.test.js
Comment (BrandonKerr) [inline test/sanityQueryService.test.js:380]: This needs to match on the ....?
also, commenting out the console logs so it doesn't make extra noise
```suggestion
        // this needs to match on the
        const expected = "['published','unlisted'] || (status == 'scheduled' && published_on >=";
        // console.log(clauses[1].condition);
        // console.log(expected)
```

File: test/sanityQueryService.test.js
Comment (BrandonKerr) [inline test/sanityQueryService.test.js:398]: I feel like this should have a unique test name. Maybe something like `withUserPermissionsForPlusUser`?

File: src/filterBuilder.js
Comment (BrandonKerr) [inline src/filterBuilder.js:28]: This would be nice to have read from the .env
```suggestion
        this.debug = process.env.DEBUG === 'true' || false;
```
(note I didn't test this suggested change)

**Review by BrandonKerr — APPROVED** [review]
I think this looks great! I left a few small notes, and would want @roxanariza's final approval for the nuances of the ContentRepository logic before merging

File: src/services/sanity.js
Comment (roxanariza) [inline src/services/sanity.js:72]: I don't know if this line of code is useful here.

**Review by roxanariza — APPROVED** [review]
The code looks good overall. I compared the logic with the RailContent implementation and noticed a few minor differences in the _applyContentStatuses method: to view future scheduled content, both getFutureScheduledContentsOnly must be set to true, and the 'scheduled' status needs to be included in the statuses. In RailContent, simply setting getFutureScheduledContentsOnly to true is enough. Your modification makes sense to me, and I agree with it.

File: src/services/sanity.js
Comment (BrandonKerr) [reply src/services/sanity.js:72]: I wondered about that too 😂 but noticed this in Adrian's table in the PR comment:

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>
<head>	
</head>

<body>

Method | Uses New Filter | User Raw Query | Hardcoded | Uses FetchAll | Has Test |  
-- | -- | -- | -- | -- | -- | --
fetchSongArtistCount | 1 |   | 1 |   | 1 |  


</body>

</html>
So I think it still needs to be updated to actually use the filter and to remove the hardcoded line below it

Comment (roxanariza) [reply src/services/sanity.js:72]: You are right. But it was strange to see that line of code there, especially since I was only focusing on the 'Uses New Filter - true' column in the table :)

Comment (adrian-musora) [reply src/services/sanity.js:72]: I had originally planned to migrate it, but I think this is used in generic pages, not in catalogues, so it doesn't make sense to filter the extra information. 
It's been removed.

File: test/sanityQueryService.test.js
Comment (adrian-musora) [reply test/sanityQueryService.test.js:380]: Updated comment.

**Pull request merged**
